File: 815. Bus Routes/815.js
// 815. Bus Routes solution in JavaScript
// https://leetcode.com/problems/bus-routes/description/
// https://leetcode.com/problems/bus-routes/solutions/2329031/fully-commented-javascript-version-of-the-most-popular-solution/

/**
 * @param {number[][]} routes
 * @param {number} source
 * @param {number} target
 * @return {number}
 */
var numBusesToDestination = function(routes, source, target) {
    let stop_routes = new Map();

    for (let route = 0; route < routes.length; route++) {
        for (let stop of routes[route]) {
            if (!stop_routes.has(stop))
                stop_routes.set(stop, new Set());

            stop_routes.get(stop).add(route);
        }
    }

    let stops_to_visit = new Queue();

    stops_to_visit.enqueue([source, 0]);

    let visited_stops = new Set();
    visited_stops.add(source);

    let visited_routes = new Array(routes.length);

    while (!stops_to_visit.isEmpty()) {
        let stop = stops_to_visit.front()[0];
        let bus = stops_to_visit.front()[1];

        stops_to_visit.dequeue();

        if (stop == target) return bus;

        for (let route of stop_routes.get(stop)) {
            if (visited_routes[route])
                continue;

            for (let stop of routes[route]) {
                if (!visited_stops.has(stop)) {
                    visited_stops.add(stop);
                    stops_to_visit.enqueue([stop, bus + 1]);
                }
            }

            visited_routes[route] = true;
        }
    }

    return -1;
};

File: 815. Bus Routes/815.go
// 815. Bus Routes solution in GoLang
// https://leetcode.com/problems/bus-routes/description/

package leetcode

func numBusesToDestination(routes [][]int, source int, target int) int {
    stop_routes := make(map[int]map[int]bool)

    for route := 0; route < len(routes); route++ {
        for _, stop := range routes[route] {
            _, is_present := stop_routes[stop]
            if !is_present {
                stop_routes[stop] = map[int]bool{}
            }

            stop_routes[stop][route] = true
        }
    }

    type Transfer struct {
        stop int
        bus int
    }
    var stops_to_visit []Transfer
    stops_to_visit = append(stops_to_visit, Transfer{source, 0})

    visited_stops := make(map[int]bool)
    visited_stops[source] = true

    visited_routes := make(map[int]bool)

    for len(stops_to_visit) > 0 {
        stop_visit := stops_to_visit[0]
        stops_to_visit = stops_to_visit[1:]

        if stop_visit.stop == target {
            return stop_visit.bus
        }

        for stop_route, _ := range stop_routes[stop_visit.stop] {
            if visited_routes[stop_route] == true {
                continue
            }

            for _, stop := range routes[stop_route] {
                _, found := visited_stops[stop]
                if !found {
                    visited_stops[stop] = true
                    stops_to_visit = append(stops_to_visit, Transfer{stop, stop_visit.bus + 1})
                }
            }

            visited_routes[stop_route] = true
        }
    }

    return -1
}